### ISSUE TITLE
ipython with matplotlib $DISPLAY issue

### ISSUE THREAD

**Issue by reconmaster** [jorgenschaefer/elpy#1099]
I've got some strange behavior with Eply using ipython as my python executable.

I have `(elpy-use-ipython)` when ipython is installed. I then have `(setq python-shell-interpreter "ipython" python-shell-interpreter-args "--simple-prompt --pprint")` to resolve the prompt issues discussed in #949 with the new ipython.

The problem is that if I try to launch ipython from any python code with C-c C-z using elpy and then try to use any of the matplotlib figures with the interactive backend ('TkAgg'), I get the follow error:

```
TclError: no display name and no $DISPLAY environmenta variable
```

I can confirm that in that ipython environement, running `!echo $DISPLAY` produces no output.

However, it is only launching ipython with elpy that causes this problem.

When I use ipython in any other way in emacs, I have no problems. Using the same configuration, if I launch ipython with `run-python` and then try using matplotlib, all my plots work fine and `!echo $DISPLAY` produces a value like `:0`.

In fact, once I have launched ipython using `run-python`, subsequently using C-c C-z to switch to that environment, I can use any of the matplotlib functionality with no problem. In fact, if I close that ipython environment and relaunch using only elpy's C-c C-z, it continues to work fine. The opposite occurs if I use elpy's C-c C-z alone and then any subsequent luanches of the ipython environement, even with run-python alone give the same error above.

I suspect that however elpy launches the ipython environment if one isn't already running is failing to take the environmental $DISPLAY into account.

Let me know if you know any additional information.

**Comment by jorgenschaefer** [comment]
Hello, and thanks for the report! That is strange indeed! :-D

Do you have `elpy-dedicated-shells` set to `t`?
What's the output of `M-: (python-shell-parse-command)`?

**Comment by reconmaster** [comment]
I had `eply-dedicated-shells` to `nil`, but after changing to `t`, I had the same problem. The output is `ipython --simple-prompt --pprint`.  For now, I am just having the --daemon do one `(run-python)` during launch so I don't forget to do it manually when it first starts.

**Comment by jorgenschaefer** [comment]
Just trying to narrow down the culprit …

`M-x elpy-switch-to-shell` creates a Python shell with the buffer
`M-x run-python` does not
Does `M-: (elpy-shell-get-or-create-process)` create a Python shell with the problem?

**Comment by reconmaster** [comment]
I've updated environments on both an osx machine and a newer version of arch. Both of those seem to behave as expected. I'm going to close this for now and assume this is not an issue with elpy. I'll provision the same environment on my work computer Mon and see if it is resolved.

I did discover that TkAgg does cause some behavior issues and seems to hang in emacs. I suppose this is probably somewhat related to this bug:

https://bugs.launchpad.net/python-mode/+bug/1103588

I have disabled TkAgg for now though this is the recommended backend on the matplotlib docs.

Some advice for fellow elpy/matplotlib users. I have found that if using matplotlib with elpy with the ipython shell with virtual environments (pyenv or virtualenvwrapper), Qt4Agg is the best bet for functionality. This is both *nix and osx systems. This does require instally pyside into the virtualenv. I did find that the MacOSX backend will work without installing pyside though the docs mention some lack in functionality.

I'll reopen this issue if I find continuing issues. Thanks again for your work on elpy.
